lib/vrl/stdlib/src/to_bool.rs

```diff
@@ -178,13 +178,15 @@ impl Expression for ToBoolFn {
     }
 
     fn type_def(&self, state: &state::Compiler) -> TypeDef {
-        TypeDef::new()
-            .with_fallibility(
-                self.value.type_def(state).has_kind(
-                    Kind::Bytes | Kind::Timestamp | Kind::Array | Kind::Object | Kind::Regex,
-                ),
-            )
-            .boolean()
+        let td = self.value.type_def(state);
+
+        TypeDef::boolean().with_fallibility(
+            td.contains_bytes()
+                || td.contains_timestamp()
+                || td.contains_array()
+                || td.contains_object()
+                || td.contains_regex(),
+        )
```

Comment:
Considering the fallible catchall in the `match` expression, I would have negated that to the known infallible cases `!(boolean || integer || float || null)`, since that could otherwise become an easy oversight if we add/remove variants on the enum variant. Since that it pretty unlikely I'm fine either way, though.